### PR TITLE
feat(Dynamic Prompt): Evaluate rules using Peer Group work

### PR DESCRIPTION
**Pull request by geoffreykwan**
## Changes

- FeedbackRuleEvaluator.getFeedbackRule() can now accept a single CRaterResponse or an array of CRaterResponses
- Dynamic prompt can figure out which dynamic prompt to display based on work and scores from the members of a Peer Chat group

## Test

- Make sure the FeedbackRuleEvaluator works when it receives one CRaterResponse (when you do not specify the dynamicPrompt.peerGroupingTag)
- Make sure the FeedbackRuleEvaluator works when it receives an array of CRaterResponses (when you specify the dynamicPrompt.peerGroupingTag)

Here is the Peer Chat component content I used for testing (when peerGroupingTag is specified).
```
{
    "id": "gkvyzuils1",
    "type": "PeerChat",
    "prompt": "",
    "showSaveButton": false,
    "showSubmitButton": false,
    "peerGroupingTag": "847zgyqp1p",
    "questionBank": [],
    "dynamicPrompt": {
        "enabled": true,
        "peerGroupingTag": "847zgyqp1p",
        "prePrompt": "This is the pre prompt",
        "postPrompt": "This is the post prompt",
        "referenceComponent": {
            "nodeId": "node49",
            "componentId": "2aabvbcukq"
        },
        "rules": [
            {
                "id": "4y3h9u0sdh",
                "expression": "isSecondToLastSubmit",
                "prompt": "Prompt when you still have a submit to use"
            },
            {
                "id": "df92gk754j",
                "expression": "2 && 3",
                "prompt": "Prompt when you got idea 2 and idea 3"
            },
            {
                "id": "7j2is1360s",
                "expression": "2 || 3",
                "prompt": "Prompt when you got idea 2 or idea 3"
            },
            {
                "id": "votbik4ncz",
                "expression": "hasKIScore(5)",
                "prompt": "Prompt when you got 5"
            },
            {
                "id": "hwp6sxf8vr",
                "expression": "hasKIScore(4)",
                "prompt": "Prompt when you got 4"
            },
            {
                "id": "swxzc0j7q0",
                "expression": "hasKIScore(3)",
                "prompt": "Prompt when you got 3"
            },
            {
                "id": "ebtgds3b1r",
                "expression": "hasKIScore(2)",
                "prompt": "Prompt when you got 2"
            },
            {
                "id": "z4yu37jcis",
                "expression": "hasKIScore(1)",
                "prompt": "Prompt when you got 1"
            },
            {
                "id": "isDefault",
                "expression": "isDefault",
                "prompt": "Default prompt"
            }
        ]
    }
}
```

#840

### CODE REVIEW

File: src/assets/wise5/components/common/feedbackRule/FeedbackRuleEvaluator.ts [hunk 1]
```diff
@@ -20,45 +20,92 @@ export class FeedbackRuleEvaluator {
...
+        this.component.hasMaxSubmitCountAndUsedAllSubmits(response.submitCounter)
+      );
+    } else {
+      return (
+        this.hasMaxSubmitAndIsFinalSubmitRule(feedbackRule) &&
+        response.some((response: CRaterResponse) => {
+          return this.component.hasMaxSubmitCountAndUsedAllSubmits(response.submitCounter);
+        })
+      );
+    }
```
Comment (hirokiterashima) [inline src/assets/wise5/components/common/feedbackRule/FeedbackRuleEvaluator.ts]: Use ternary operator to reduce duplication and complexity
```suggestion
    return (
      this.hasMaxSubmitAndIsFinalSubmitRule(feedbackRule) &&
      (response instanceof CRaterResponse
        ? this.component.hasMaxSubmitCountAndUsedAllSubmits(response.submitCounter)
        : response.some((response: CRaterResponse) => {
            return this.component.hasMaxSubmitCountAndUsedAllSubmits(response.submitCounter);
          }))
    );
```

File: src/assets/wise5/components/common/feedbackRule/FeedbackRuleEvaluator.ts [hunk 2]
```diff
@@ -20,45 +20,92 @@ export class FeedbackRuleEvaluator {
...
+        this.isSecondToLastSubmit(response.submitCounter)
+      );
+    } else {
+      return (
+        this.hasMaxSubmitAndIsSecondToLastSubmitRule(feedbackRule) &&
+        response.some((response: CRaterResponse) => {
+          return this.isSecondToLastSubmit(response.submitCounter);
+        })
+      );
+    }
```
Comment (hirokiterashima) [inline src/assets/wise5/components/common/feedbackRule/FeedbackRuleEvaluator.ts]: Use ternary operator to reduce duplication and complexity
```suggestion
    return (
      this.hasMaxSubmitAndIsSecondToLastSubmitRule(feedbackRule) &&
      (response instanceof CRaterResponse
        ? this.isSecondToLastSubmit(response.submitCounter)
        : response.some((response: CRaterResponse) => {
            return this.isSecondToLastSubmit(response.submitCounter);
          }))
    );
```

File: src/assets/wise5/components/common/feedbackRule/FeedbackRuleEvaluator.ts [hunk 3]
```diff
@@ -20,45 +20,92 @@ export class FeedbackRuleEvaluator {
...
+    if (response instanceof CRaterResponse) {
+      return feedbackRule.expression === 'isNonScorable' && response.isNonScorable();
+    } else {
+      return (
+        feedbackRule.expression === 'isNonScorable' &&
+        response.some((response: CRaterResponse) => {
+          return response.isNonScorable();
+        })
+      );
+    }
```
Comment (hirokiterashima) [inline src/assets/wise5/components/common/feedbackRule/FeedbackRuleEvaluator.ts]: Use ternary operator to reduce duplication and complexity

File: src/assets/wise5/components/common/feedbackRule/FeedbackRuleEvaluator.ts [hunk 4]
```diff
@@ -128,7 +185,13 @@ export class FeedbackRuleEvaluator {
...
       evaluator = new IdeaTermEvaluator(term);
     }
-    return evaluator.evaluate(response);
+    if (response instanceof CRaterResponse) {
+      return evaluator.evaluate(response);
+    } else {
+      return response.some((response: CRaterResponse) => {
+        return evaluator.evaluate(response);
+      });
+    }
```
Comment (hirokiterashima) [inline src/assets/wise5/components/common/feedbackRule/FeedbackRuleEvaluator.ts]: Use ternary operator to reduce duplication and complexity

**Review by hirokiterashima — CHANGES_REQUESTED** [review]
Functionality works well. I wrote some inline code comment suggestions.

**Comment by geoffreykwan** [comment]
Changing those sections to ternary make them a lot harder to read.

**Comment by hirokiterashima** [comment]
To me the changes don't make the code a lot harder to read. I'm reading less code and the logic is easy to follow. Why do you find it harder to read? We can also ask @breity what he thinks.

The issue I have with the code is the duplication and complexity that they add. The code becomes longer and more repetitive. I think ternary conditional solves those issues. Can you suggest another way to write the code that avoids those issues?

**Review by hirokiterashima — APPROVED** [review]
LGTM

**Pull request merged**